### PR TITLE
RavenDB-4902 - ready to merge - separated the work to a new session implementation

### PR DESCRIPTION
**Pull request by fitzchak**

### CODE REVIEW

File: src/Raven.Client/Document/DocumentStore.cs [hunk 1]
```diff
@@ -278,11 +278,10 @@ public override IDocumentSession OpenSession(OpenSessionOptions options)
             currentSessionId = sessionId;
             try
             {
-                var session = new DocumentSession(options.Database, this, Listeners, sessionId,
-                    SetupCommands(DatabaseCommands, options.Database, options.Credentials, options))
-                    {
-                        DatabaseName = options.Database ?? DefaultDatabase ?? MultiDatabase.GetDatabaseName(Url)
-                    };
+                var databaseName = options.Database ?? DefaultDatabase ?? MultiDatabase.GetDatabaseName(Url);
+                var requestExecuter = _requestExecuters.GetOrAdd(databaseName, new RequestExecuter(Url, databaseName, ApiKey));
```
Comment (ayende) [inline src/Raven.Client/Document/DocumentStore.cs]: Here you are creating a request executer every single time.
Better to have Lazy<RequestExecuter> and create that via a delegate

File: src/Raven.Client/Document/SessionOperations/NewLoadOperation.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+            _idsToCheckOnServer.Add(id);
+        }
+
+        public void ByIds(IEnumerable<string> ids)
+        {
+            _ids = ids.ToArray();
+            foreach (var id in _ids
+                .Distinct(StringComparer.OrdinalIgnoreCase)
+                .ToArray())
```
Comment (ayende) [inline src/Raven.Client/Document/SessionOperations/NewLoadOperation.cs]: Why do you call ToArray here?
And _ids can be a HashSet, instead

File: src/Raven.Client/Document/SessionOperations/NewLoadOperation.cs [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+            }
+            return (T) entity;
+        }
+
+        public T[] GetDocuments<T>()
+        {
+            var finalResults = new T[_ids.Length];
+            for (int i = 0; i < _ids.Length; i++)
+            {
+                var id = _ids[i];
```
Comment (ayende) [inline src/Raven.Client/Document/SessionOperations/NewLoadOperation.cs]: why do you have this line?

File: src/Raven.Client/Document/InMemoryDocumentSessionOperations.cs [hunk 4]
```diff
@@ -1137,6 +1188,7 @@ public void ExplicitlyVersion(object entity)
         /// </summary>
         public virtual void Dispose()
         {
+            _releaseOperationContext.Dispose();
```
Comment (ayende) [inline src/Raven.Client/Document/InMemoryDocumentSessionOperations.cs]: Also need a finalizer to handle this

File: src/Raven.Client/Document/DocumentStore.cs (hunk 1 above)
Comment (fitzchak) [reply src/Raven.Client/Document/DocumentStore.cs]: It is once per database, not every single time.

File: src/Raven.Client/Document/SessionOperations/NewLoadOperation.cs (hunk 2 above)
Comment (fitzchak) [reply src/Raven.Client/Document/SessionOperations/NewLoadOperation.cs]: I'll use SortedSet has I need to get the position of an id.

File: src/Raven.Client/Document/SessionOperations/NewLoadOperation.cs (hunk 3 above)
Comment (fitzchak) [reply src/Raven.Client/Document/SessionOperations/NewLoadOperation.cs]: I'm using the id to pass to GetDocument method. I'm not sure what is the issue you see here.

File: src/Raven.Client/Document/Async/AsyncDocumentSession.cs [hunk 5]
```diff
@@ -712,44 +710,32 @@ public Task<T[]> LoadAsync<T>(IEnumerable<ValueType> ids, CancellationToken toke
...
+            {
+                await RequestExecuter.ExecuteAsync(command, Context);
+                loadOeration.SetResult(command.Result);
             }
-        }
 
-        public Task<T[]> LoadAsync<T>(IEnumerable<string> ids, CancellationToken token = default(CancellationToken))
-        {
-            return LoadAsyncInternal<T>(ids.ToArray(), token);
+            return loadOeration.GetDocuments<T>();
```
Comment (ppekrol) [inline src/Raven.Client/Document/Async/AsyncDocumentSession.cs]: what will be the order here? it must be the same as specified in the LoadAsync ids parameter and IIRC it must support duplicates

Comment (fitzchak) [reply src/Raven.Client/Document/Async/AsyncDocumentSession.cs]: @ppekrol You're are right, this is why we cannot use SortedSet for the internal ids so I'm using a regular array instead.

**Comment by fitzchak** [comment]
All tests are passing

**Pull request merged**